## <font color=red>15596</font>

In [ ]:
solve=sum

In [ ]:
def solve(a):
    result = 0
    for A in a:
        result += A
    return result

## <font color=red>4673</font>

In [ ]:
Range=range(10)
List=set([1001*a+101*b+11*c+2*d for a in Range for b in Range for c in Range for d in Range])
for i in range(10001):
    if i not in List:
        print(i)

## <font color=red>1065</font>
어떤 양의 정수 X의 각 자리가 등차수열을 이룬다면, 그 수를 한수라고 한다. 등차수열은 연속된 두 개의 수의 차이가 일정한 수열을 말한다. N이 주어졌을 때, 1보다 크거나 같고, N보다 작거나 같은 한수의 개수를 출력하는 프로그램을 작성하시오. 

In [31]:
N=int(input())
LIST=[i for i in range(1,100)]
for i in range(10):
    LIST.append(100*i+10*i+i)
    for j in range(1,10):
        if i+2*j<10:
            LIST.append(100*i+10*(i+j)+(i+2*j))
        if i-2*j>=0:
            LIST.append(100*i+10*(i-j)+(i-2*j))
score=0
for i in range(1,N+1):
    if i in LIST:
        score+=1
print(score)

210
105


In [30]:
N=int(input())
LIST=[i for i in range(1,100)] + [100*i+10*i+i for i in range(10)] + [100*i+10*(i+j)+(i+2*j) for i in range(10) for j in range(1,10) if i+2*j<10] + [100*i+10*(i-j)+(i-2*j) for i in range(10) for j in range(1,10) if i-2*j>=0]
score=0
for i in range(1,N+1):
    if i in LIST:
        score+=1
print(score)

1000
144


In [6]:
print(sum(i//100+i%10==i//10%10*2 or i<100 for i in range(1,int(input())+1)))

110
99


## <font color=red>1065</font> 변형문제
위의 문제를 약간 바꾸어보자
양의 정수 X의 각 자리수들을 적절하게 재배열하여 등차수열을 이룰 수 있다면 X를 `두수`라고 하자.  
자연수 N이 주어졌을 때, 1보다 크거나 같고 N보다 작거나 같은 `두수`의 개수를 출력하는 프로그램을 작성하시오. 
 
예를 들면,
- 110을 입력했을 때 100이 출력되고
- 210을 입력했을 때 113이 출력되고 
- 1000을 입력했을 때 220이 출력된다.

In [3]:
# itertools 모듈의 permutations 메서드를 이용해서 풀면 편하다.
import itertools

pool = ['A', 'B', 'C']
print(list(map(''.join, itertools.permutations(pool)))) # 3개의 원소로 수열 만들기

['ABC', 'ACB', 'BAC', 'BCA', 'CAB', 'CBA']


In [38]:
import itertools

N=int(input())
LIST=[]
for i in range(10):
    LIST.append([f"{i}",f"{i}",f"{i}"])
    LIST.append([f"{i}",f"{i}"])
    LIST.append([f"{i}"])
    for j in range(1,10):
        if i+2*j<10:
            LIST.append([f"{i}",f"{i+j}",f"{i+2*j}"])
        if i+j<10:
            LIST.append([f"{i}", f"{i+j}"])

X=set()
for List in LIST:
    X.update( list(set(map(''.join, itertools.permutations(List))))  )
X = set(map(int,X))
X = list(X)
score=0
for i in range(1,N+1):
    if i in X:
        score+=1
print(score)

1000
220
